### PR TITLE
[Guilds/Core] Allow guilds with different GuildTypes to have the same name

### PR DESCRIPTION
**Pull request by StratSus**
You should now be able to create a Guild, Circle and Community with the same name.

### CODE REVIEW

File: Source/NexusForever.WorldServer/Game/Guild/GlobalGuildManager.cs [hunk 1]
```diff
@@ -399,5 +400,27 @@ public void HandleGuildOperation(Player player, ClientGuildOperation operation)
                 return info;
             }
         }
+        private class GuildNameEqualityComparer : IEqualityComparer<(GuildType, string)>
+        {
+            public bool Equals((GuildType, string) x, (GuildType, string) y)
+            {
+                if (x.Item1 == y.Item1)
+                {
+                    if (ReferenceEquals(x.Item2, y.Item2)) return true;
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Game/Guild/GlobalGuildManager.cs]: To follow code style in the rest of the project return should be on a separate line.

File: Source/NexusForever.WorldServer/Game/Guild/GlobalGuildManager.cs [hunk 2]
```diff
@@ -399,5 +400,27 @@ public void HandleGuildOperation(Player player, ClientGuildOperation operation)
                 return info;
             }
         }
+        private class GuildNameEqualityComparer : IEqualityComparer<(GuildType, string)>
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Game/Guild/GlobalGuildManager.cs]: Might be better to move the comparer to a separate file to reduce clutter in a already large class.

File: Source/NexusForever.WorldServer/Game/Guild/GlobalGuildManager.cs [hunk 3]
```diff
@@ -399,5 +400,27 @@ public void HandleGuildOperation(Player player, ClientGuildOperation operation)
...
+        {
+            public bool Equals((GuildType, string) x, (GuildType, string) y)
```
Comment (Rawaho) [inline Source/NexusForever.WorldServer/Game/Guild/GlobalGuildManager.cs]: To follow code style of the rest of the project, tuple members should be named or deconstructed.

**Review by Rawaho — CHANGES_REQUESTED** [review]
Just a few code style changes but functionally looks good.

File: Source/NexusForever.WorldServer/Game/Guild/GlobalGuildManager.cs (hunk 3 above)
Comment (StratSus) [reply Source/NexusForever.WorldServer/Game/Guild/GlobalGuildManager.cs]: When I try to name the tuple members, I get an error stating that the tuple names need to be the same as those in the Interface method, which I don't think I can change.

File: Source/NexusForever.WorldServer/Game/Guild/GuildNameEqualityComparer.cs [hunk 4]
```diff
@@ -0,0 +1,39 @@
+﻿using NexusForever.WorldServer.Game.Guild.Static;
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Text;
+using System.Threading.Tasks;
```
Comment (MaxtorCoder) [inline Source/NexusForever.WorldServer/Game/Guild/GuildNameEqualityComparer.cs]: Remove any unused directives here

File: Source/NexusForever.WorldServer/Game/Guild/GuildNameEqualityComparer.cs [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+namespace NexusForever.WorldServer.Game.Guild
+{
+    internal class GuildNameEqualityComparer : IEqualityComparer<(GuildType, string)>
```
Comment (MaxtorCoder) [inline Source/NexusForever.WorldServer/Game/Guild/GuildNameEqualityComparer.cs]:
```suggestion
    internal class GuildNameEqualityComparer : IEqualityComparer<(GuildType Type, string Name)>
```

File: Source/NexusForever.WorldServer/Game/Guild/GuildNameEqualityComparer.cs [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+                    return false;
+
+                if (y.Item2 is null)
+                    return false;
+
+                if (x.Item2.GetType() != y.Item2.GetType())
+                    return false;
+
+                return string.Equals(x.Item2, y.Item2, StringComparison.InvariantCultureIgnoreCase);
+            }
```
Comment (MaxtorCoder) [inline Source/NexusForever.WorldServer/Game/Guild/GuildNameEqualityComparer.cs]:
```suggestion
        public bool Equals((GuildType Type, string Name) x, (GuildType Type, string Name) y)
        {
            if (x.Type == y.Type)
            {
                if (ReferenceEquals(x.Name, y.Name))
                    return true;

                if (x.Name is null)
                    return false;

                if (y.Name is null)
                    return false;

                if (x.Name.GetType() != y.Name.GetType())
                    return false;

                return string.Equals(x.Name, y.Name, StringComparison.InvariantCultureIgnoreCase);
            }
```
 This should work with naming tuple members

File: Source/NexusForever.WorldServer/Game/Guild/GlobalGuildManager.cs [hunk 7]
```diff
@@ -35,8 +36,8 @@ public sealed class GlobalGuildManager : Singleton<GlobalGuildManager>
         private ulong nextGuildId;
 
         private readonly Dictionary</*guildId*/ ulong, GuildBase> guilds = new();
-        private readonly Dictionary<string, ulong> guildNameCache = new(StringComparer.InvariantCultureIgnoreCase);
-        private readonly Dictionary<ulong, List<ulong>> guildMemberCache = new();
+        private readonly Dictionary<(GuildType, string), ulong> guildNameCache = new(new GuildNameEqualityComparer());
```
Comment (MaxtorCoder) [inline Source/NexusForever.WorldServer/Game/Guild/GlobalGuildManager.cs]: I guess name the tuple members here as well?

**Review by MaxtorCoder — CHANGES_REQUESTED** [review]

File: Source/NexusForever.WorldServer/Game/Guild/GuildNameEqualityComparer.cs (hunk 6 above)
Comment (StratSus) [reply Source/NexusForever.WorldServer/Game/Guild/GuildNameEqualityComparer.cs]: Thanks. Brainfart.

**Review by Rawaho — APPROVED** [review]

**Pull request merged**